File: build_snapshot.ts
import { config } from "dotenv";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
config({ path: resolve(__dirname, ".env") });

import { Daytona, Image } from "@daytona/sdk";
import type { CreateSandboxFromSnapshotParams } from "@daytona/sdk";

type Sandbox = Awaited<ReturnType<Daytona["create"]>>;

const OPENCODE_REF = process.argv[2] || "dev";
const SNAPSHOT_PREFIX = process.argv[3] || "opencode-server";

function buildSnapshotName(): string {
  const safeRef = OPENCODE_REF.replaceAll("/", "-");
  const ts = new Date().toISOString().replace(/[-:T]/g, "").slice(0, 15);
  const suffix = Math.floor(1000 + Math.random() * 9000);
  return `${SNAPSHOT_PREFIX}-${safeRef}-${ts}-${suffix}`;
}

function elapsed(t0: number): string {
  return `${((performance.now() - t0) / 1000).toFixed(1)}s`;
}

function env(key: string): string {
  const val = process.env[key] ?? "";
  if (!val) console.log(`    WARNING: ${key} not set in .env`);
  return val;
}

async function checkServe(sandbox: Sandbox): Promise<boolean> {
  const result = await sandbox.process.executeCommand(
    "cd /workspace && opencode serve --port 4096 --hostname 127.0.0.1 > /tmp/serve.log 2>&1 &" +
      " sleep 4 && curl -sf http://127.0.0.1:4096/global/health",
    undefined,
    undefined,
    30,
  );
  console.log(`    serve health check exit code: ${result.exitCode}`);
  const body = result.result.trim();
  if (body) {
    console.log(`    /global/health: ${body.slice(0, 300)}`);
    return true;
  }
  const logs = await sandbox.process.executeCommand("cat /tmp/serve.log", undefined, undefined, 10);
  console.log(`    WARNING: no response. Server logs:\n${logs.result.trim().slice(0, 500)}`);
  return false;
}

async function main() {
  const daytona = new Daytona();

  // --- Phase 1: Create snapshot ---
  const snapshotName = buildSnapshotName();
  const dockerfilePath = resolve(__dirname, "Dockerfile");

  console.log(`=== Creating snapshot: ${snapshotName} ===`);
  console.log(`    Dockerfile: ${dockerfilePath}`);
  console.log(`    Ref:        ${OPENCODE_REF}`);
  console.log();

  const image = Image.fromDockerfile(dockerfilePath);
  (image as any)._dockerfile = (image as any)._dockerfile.replace(
    "ARG OPENCODE_REF=dev",
    `ARG OPENCODE_REF=${OPENCODE_REF}`,
  );

  let t0 = performance.now();
  const snapshot = await daytona.snapshot.create(
    {
      name: snapshotName,
      image,
      resources: { cpu: 2, memory: 4, disk: 10 },
    },
    { onLogs: (chunk: string) => process.stdout.write(`  [build] ${chunk}`) },
  );
  console.log(`\n=== Snapshot created: ${snapshot.name} (${elapsed(t0)}) ===\n`);

  // --- Phase 2: Create sandbox ---
  console.log("=== Creating sandbox from snapshot ===");
  t0 = performance.now();

  const sandbox = await daytona.create(
    {
      snapshot: snapshot.name,
      envVars: { OPENAI_API_KEY: env("OPENAI_API_KEY") },
      autoStopInterval: 0,
    } satisfies CreateSandboxFromSnapshotParams,
    { timeout: 120 },
  );
  console.log(`    Sandbox ID: ${sandbox.id} (${elapsed(t0)})`);

  // --- Phase 3: Verify ---
  console.log("\n=== Verifying opencode inside sandbox ===");

  let result = await sandbox.process.executeCommand("opencode --version", undefined, undefined, 30);
  console.log(`    opencode --version: ${result.result.trim()}`);
  console.log(`    exit code: ${result.exitCode}`);

  if (result.exitCode !== 0) {
    console.log("    FAILED: opencode binary not working");
    await sandbox.delete();
    return;
  }

  if (!(await checkServe(sandbox))) {
    await sandbox.delete();
    return;
  }

  // --- Phase 4: Stop → Start → Re-verify ---
  console.log("\n=== Stopping sandbox ===");
  t0 = performance.now();
  await sandbox.stop(60);
  console.log(`    Stopped (${elapsed(t0)})`);

  console.log("\n=== Starting sandbox ===");
  t0 = performance.now();
  await sandbox.start(60);
  console.log(`    Started (${elapsed(t0)})`);

  console.log("\n=== Re-verifying after restart ===");
  result = await sandbox.process.executeCommand("opencode --version", undefined, undefined, 30);
  console.log(`    opencode --version: ${result.result.trim()}`);

  if (!(await checkServe(sandbox))) {
    await sandbox.delete();
    return;
  }

  // --- Cleanup ---
  console.log("\n=== Cleaning up sandbox ===");
  await sandbox.delete();
  console.log(`    Sandbox ${sandbox.id} deleted`);

  console.log(`\n=== Done ===`);
  console.log(`    Snapshot ready: ${snapshot.name}`);
  console.log(`    Create a sandbox with:`);
  console.log(`      daytona create --snapshot ${snapshot.name}`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: build_snapshot.py
"""Build an opencode snapshot from Dockerfile, then create a sandbox to verify it works."""

import sys
import time
import random
from pathlib import Path
from datetime import datetime, timezone

from dotenv import load_dotenv

load_dotenv(Path(__file__).parent / ".env")

from daytona import (
    Daytona,
    DaytonaConfig,
    CreateSnapshotParams,
    CreateSandboxFromSnapshotParams,
    Image,
    Resources,
)

OPENCODE_REF = sys.argv[1] if len(sys.argv) > 1 else "dev"
SNAPSHOT_PREFIX = sys.argv[2] if len(sys.argv) > 2 else "opencode-server"


def build_snapshot_name() -> str:
    safe_ref = OPENCODE_REF.replace("/", "-")
    ts = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
    suffix = random.randint(1000, 9999)
    return f"{SNAPSHOT_PREFIX}-{safe_ref}-{ts}-{suffix}"


def main():
    daytona = Daytona(DaytonaConfig())

    # --- Phase 1: Create snapshot ---
    snapshot_name = build_snapshot_name()
    dockerfile_path = Path(__file__).parent / "Dockerfile"

    print(f"=== Creating snapshot: {snapshot_name} ===")
    print(f"    Dockerfile: {dockerfile_path}")
    print(f"    Ref:        {OPENCODE_REF}")
    print()

    image = Image.from_dockerfile(str(dockerfile_path))
    t0 = time.perf_counter()

    snapshot = daytona.snapshot.create(
        CreateSnapshotParams(
            name=snapshot_name,
            image=image,
            resources=Resources(cpu=2, memory=4, disk=10),
        ),
        on_logs=lambda line: print(f"  [build] {line}", end=""),
    )

    build_secs = time.perf_counter() - t0
    print(f"\n=== Snapshot created: {snapshot.name} ({build_secs:.1f}s) ===\n")

    # --- Phase 2: Create sandbox from snapshot ---
    print("=== Creating sandbox from snapshot ===")
    t0 = time.perf_counter()

    sandbox = daytona.create(
        CreateSandboxFromSnapshotParams(
            snapshot=snapshot.name,
            env_vars={"OPENAI_API_KEY": _get_env("OPENAI_API_KEY")},
            auto_stop_interval=0,
        ),
        timeout=120,
    )

    create_secs = time.perf_counter() - t0
    print(f"    Sandbox ID: {sandbox.id} ({create_secs:.1f}s)")

    # --- Phase 3: Verify ---
    print("\n=== Verifying opencode inside sandbox ===")

    result = sandbox.process.exec("opencode --version", timeout=30)
    print(f"    opencode --version: {result.result.strip()}")
    print(f"    exit code: {result.exit_code}")

    if result.exit_code != 0:
        print("    FAILED: opencode binary not working")
        sandbox.delete()
        return

    result = sandbox.process.exec(
        "opencode serve --port 4096 --hostname 127.0.0.1 &"
        " sleep 3 && curl -sf http://127.0.0.1:4096/doc | head -c 200",
        timeout=30,
    )
    print(f"    serve check exit code: {result.exit_code}")
    if result.result.strip():
        print(f"    /doc response (first 200 chars): {result.result.strip()[:200]}")

    # --- Cleanup ---
    print("\n=== Cleaning up sandbox ===")
    sandbox.delete()
    print(f"    Sandbox {sandbox.id} deleted")

    print(f"\n=== Done ===")
    print(f"    Snapshot ready: {snapshot.name}")
    print(f"    Create a sandbox with:")
    print(f'      daytona create --snapshot {snapshot.name}')


def _get_env(key: str) -> str:
    import os
    val = os.environ.get(key, "")
    if not val:
        print(f"    WARNING: {key} not set in .env")
    return val


if __name__ == "__main__":
    main()
